File: Problems/0005. Longest Palindromic Substring/solution2.py
# DP Solution
class Solution:
    def longestPalindrome(self, s: str) -> str:
        result = ""

        dp = [[False] * len(s) for i in range(len(s))]

        for i in range(len(s) - 1, -1, -1):
            for j in range(i, len(s)):
                if s[i] == s[j] and (j - i + 1 <= 2 or dp[i + 1][j - 1]):
                    dp[i][j] = True
                    if j - i + 1 > len(result):
                        result = s[i:j + 1]
        return result

File: Problems/0647. Palindromic Substrings/solution2.java
# DP Solution
class Solution:
    def countSubstrings(self, s: str) -> int:
        result = 0

        dp = [[False] * len(s) for i in range(len(s))]

        for row in range(len(s) - 1, -1, -1):
            for col in range(row, len(s)):
                if s[row] == s[col] and (col - row < 2 or dp[row + 1][col - 1]):
                    dp[row][col] = True
                    result += 1

        return result
